MyEtherWallet/MyEtherWallet
# Electron App

Begins addressing #749. This PR is a first pass at an Electron app and a way of updating via Github releases. As a part of implementing it, I also refactored our Webpack configs and build process to better accommodate the growing number of configs and builds we have. This PR should be considered as three discrete changes, and if people feel it would be useful, I could break it into 3 PRs:

#### (Note: Node v9 is required for building).

# What This Does

### Webpack Configs / Build Process

* Everything in `webpack_config/` and `package.json`
* Most webpack configs now run through a `makeConfig` function, allowing for more complex sharing and conditional resolving
* All webpack utilities are now in a single `config.js` file (No more `utils.js`)
* Freezer and derivation checker do not use `makeConfig`, and do not inherit from other configs. They are stand-alone, and should not break so easily if we manipulate other configs.
* Builds now get output into separate folders in `dist/`. This should allow for parallel builds in the future, and make it more clear which build you're running. Re-running a build _only_ deletes the folder it's in, not all of `dist/`.
    * `npm run build` goes into `dist/prod`
    * `npm run build:downloadable` goes into `dist/download`
    * `npm run build:electron` goes into `dist/electron-js` and `dist/electron-builds` for the Javascript it outputs, and the executables it makes respectively
        * If you want a faster build on your own platform, or your platform doesn't support building multiple platforms, there's also `npm run build:electron:[osx|windows|linux]`
* A few more environment variables have been added or changed:
    * `BUILD_DOWNLOADABLE` is true for both the HTML or Electron builds
    * `BUILD_HTML` is true only for the HTML downloadable build
    * `BUILD_ELECTRON` is true only for the Electron build
        * Electron also gets `process.platform = "darwin|win32|linux"`
* Converted @skubakdj's postBuild script into a webpack plugin

### Electron Core Code

* A new top-level folder has been added, `electron/`
* It uses node-friendly ES6 code only right now
    * This could likely get changed to use Typescript, but for dev purposes, I couldn't get this working in a timely manner. `electron-webpack` may be a solution for the dev side.
* Electron can be run in developer mode using `npm run dev:electron`
* It primarily builds `main.js`, Electron's main thread
* If / when we come up with use cases, we could also build renderer side-threads to do things like run a node for the user or other background processes.
* Electron runs in a sandbox with no direct Node integration, which disconnects the main JS bundle from node. This reduces our risk of exposing the user's system.
* Connections to node have to be explicitly made in `preload.js`, by providing functions to the window object of the main JS bundle. This is the best way that I could find to do this, but let me know if you know of a better way.
* Code is signed using the following instructions: https://www.electron.build/code-signing

### Electron Web Integrations

* The remaining changes in `common/`
* Some of the client code attempts to integrate with Electron, by either using exposed functions, or checking `process.env.BUILD_ELECTRON`
* The site version is now clickable when there's an update available, used to kick off updates to the Electron app
    * Right now this solely renders update information using Electron's `autoUpdater`, but I'd like users to be able to read patch notes directly through the site on all platforms.
* The first pass at a TitleBar was made, but put on the backburner. Though I left some of the component implementation in so that it could be quickly picked up for a Pt 2.
* NewTabLinks open links in the user's default browser, rather than within the Electron app.

# TODOs

* Ledger and Trezor don't currently work for various reasons, ideally we should be able to get these up and running
* Updates currently download _the whole app_ again. We should investigate some way of doing smaller partial updates.
* Typescript and better dev environment for Electron code.
* Offer the patch notes modal to all platforms in a non-electron reliant way. Potentially have it show older versions as well.
* We'll probably want to implement some custom chrome ala Slack or Discord, right now it just uses the default shell which looks kinda ugly

# How To Test

* Install the new dependencies
* Run Electron in development mode with `npm run dev:electron` and make sure it works
    * If you would like to test what an update looks like, flip`shouldMockUpdate` to true in `electron/updater.js`. It's not a real one, just a mock one.
* Run a build of Electron with `npm run build:electron` and make sure it works
    * If you're having trouble, try just building your platform with `npm run build:electron:[osx|windows|linux]
* Run a production build with `npm run build` and make sure nothing has changed from that build
* Run a downloadable build with `npm run build:downloadabl` and make sure nothing has changed from that build

# Screenshots

### The App
<img width="1332" alt="screen shot 2018-01-17 at 3 25 00 pm" src="https://user-images.githubusercontent.com/649992/35066376-64ce7826-fb9e-11e7-81c4-6514f1663249.png">

### Update Available
![update-available](https://user-images.githubusercontent.com/649992/35066396-6f0be9ae-fb9e-11e7-8a94-ee0a17190273.gif)

### The Update Modal
<img width="742" alt="screen shot 2018-01-17 at 3 25 19 pm" src="https://user-images.githubusercontent.com/649992/35066407-769254b0-fb9e-11e7-8678-e1506cabe9a9.png">


### Downloading...
<img width="762" alt="screen shot 2018-01-17 at 3 25 24 pm" src="https://user-images.githubusercontent.com/649992/35066420-7f2db380-fb9e-11e7-8677-68c2c802e692.png">
